tests/30rooms/13guestaccess.pl

```diff
@@ -485,6 +485,66 @@
       });
    };
 
+test "Guest users can accept invites to private rooms over federation",
+   requires => [ remote_user_fixture(), guest_user_fixture() ],
+
+   bug => "synapse#2065",
+
+   do => sub {
+      my ( $remote_user, $local_guest ) = @_;
+
+      my ( $room_id );
+
+      matrix_create_room( $remote_user )->then( sub {
+         ( $room_id ) = @_;
+
+         matrix_put_room_state( $remote_user, $room_id,
+            type      => "m.room.join_rules",
+            state_key => "",
+            content   => {
+               join_rule => "invite",
+            }
+         )
+      })->then( sub {
+         matrix_set_room_guest_access( $remote_user, $room_id, "can_join" )
+      })->then( sub {
+         matrix_invite_user_to_room( $remote_user, $local_guest, $room_id )
+      })->then( sub {
+         matrix_join_room( $local_guest, $room_id );
+      })->then( sub {
+         Future->done( 1 );
+      });
+   };
+
+test "Guest users denied access over federation if guest access prohibited",
+   requires => [ remote_user_fixture(), guest_user_fixture() ],
+
+   do => sub {
+      my ( $remote_user, $local_guest ) = @_;
+
+      my ( $room_id );
+
+      matrix_create_room( $remote_user )->then( sub {
+         ( $room_id ) = @_;
+
+         matrix_put_room_state( $remote_user, $room_id,
+            type      => "m.room.join_rules",
+            state_key => "",
+            content   => {
+               join_rule => "invite",
+            }
+         )
+      })->then( sub {
+         matrix_set_room_guest_access( $remote_user, $room_id, "forbidden" )
+      })->then( sub {
+         matrix_invite_user_to_room( $remote_user, $local_guest, $room_id )
+      })->then( sub {
+         matrix_join_room( $local_guest, $room_id )->main::expect_http_403
```

Comment:
On a purely stylistic note, I think it's becoming usual in tests to wrap the `->main::expect_http_403` onto a new line anyway, even though it does fit here, just to draw more attention to it. Its presence entirely inverts the expected logic here so it's quite important to see it.

    matrix_join_room( ... )
       ->main::expect_http_403